tests-integration/src/tests/instance_kafka_wal_test.rs

```diff
@@ -74,24 +82,196 @@ async fn test_create_database_and_insert_query(instance: Option<Box<dyn Rebuilda
     }
 }
 
-async fn execute_sql(instance: &Arc<Instance>, sql: &str) -> Output {
-    execute_sql_with(instance, sql, QueryContext::arc()).await
+struct Table {
+    name: String,
+    logical_timer: AtomicU64,
+    inserted: Mutex<Vec<u64>>,
 }
 
-async fn try_execute_sql_with(
-    instance: &Arc<Instance>,
-    sql: &str,
-    query_ctx: QueryContextRef,
-) -> Result<Output> {
-    instance.do_query(sql, query_ctx).await.remove(0)
+#[apply(both_instances_cases_with_kafka_wal)]
+async fn test_replay(rebuildable_instance: Option<Box<dyn RebuildableMockInstance>>) {
+    let Some(mut rebuildable_instance) = rebuildable_instance else {
+        return;
+    };
+    let instance = rebuildable_instance.frontend();
+
+    let output = execute_sql(&instance, "create database test").await;
+    assert_matches!(output, Output::AffectedRows(1));
+
+    let tables = create_tables("test_replay", &instance, 3).await;
+    insert_data(&tables, &instance, 5).await;
+    ensure_data_exists(&tables, &instance).await;
+
+    // Rebuilds to emulate restart which then triggers a replay.
+    let instance = rebuildable_instance.rebuild().await;
+    ensure_data_exists(&tables, &instance).await;
 }
 
-async fn execute_sql_with(
+#[apply(both_instances_cases_with_kafka_wal)]
+async fn test_flush_then_replay(rebuildable_instance: Option<Box<dyn RebuildableMockInstance>>) {
+    let Some(mut rebuildable_instance) = rebuildable_instance else {
+        return;
+    };
+    let instance = rebuildable_instance.frontend();
+
+    let output = execute_sql(&instance, "create database test").await;
+    assert_matches!(output, Output::AffectedRows(1));
+
+    let tables = create_tables("test_flush_then_replay", &instance, 3).await;
```

Comment:
How about adding 10 tables? In the IT, we configure 5 topics by default. If the number of tables is too few might ignore bugs.